File: src/Tile/TileSet.m
classdef TileSet < handle

    properties
        TileTypes = [];
        Concs = [];
        Kd = 1;
        concs_max = 0;
        usetileconc = 0;
    end

    methods
        function obj = TileSet(mode, usetileconc, Kd, concs_max)
            if nargin == 0
                msg = "TileSet needs mode and concentrations.";
                error(msg)
            elseif nargin > 0
                % Populate a dictionary of Tile Types and their
                % Concentrations according to conflict rules
                if mode == "linear"
                    [obj.TileTypes, obj.Concs] = obj.init_linear_tile_set(concs_max);
                elseif mode == "radial"
                    [obj.TileTypes, obj.Concs] = obj.init_radial_tile_set(concs_max);
                else
                    error("Invalid growth mode.")
                end
                obj.concs_max = concs_max;
                obj.Kd = Kd;
                obj.usetileconc = usetileconc;
            end 
        end

        %% Tile conc modifiers
        function dec_conc(obj, defn)
            % Decreases the concentration of a tile type by 1 unit
            index = ismember(obj.TileTypes, defn, 'rows');
            idx = find(index);
            if isempty(idx)
                error("Incorrect tile definition.")
            end
            if obj.Concs(idx) > 0
                obj.Concs(idx) = obj.Concs(idx) - 1;
            end
        end

        function inc_conc(obj, defn)
            % Increases the concentration of a tile type by 1 unit
            index = ismember(obj.TileTypes, defn, 'rows');
            idx = find(index);
            if isempty(idx)
                error("Incorrect tile definition.")
            end
            obj.Concs(idx) = obj.Concs(idx) + 1;
        end

        function buffer(obj, num)
            % Increases the concentration of all tile types by 'num' units
            obj.Concs = obj.Concs + num;
        end

        function apply_branching_factor(obj, b)
            % Re-weights branching tile types according to branching factor
            for tt = 1:size(obj.TileTypes, 1)
                % A branching tile has additional output glues defined on its 4 and 6 elements
                if obj.TileTypes(tt, 4) ~= "NULL" && obj.TileTypes(tt, 6) ~= "NULL"
                    obj.Concs(tt) = obj.Concs(tt) * b;
                end
            end
        end

        function apply_turning_factor(obj, t)
            % Re-weights turning tile types according to turning factor
            for tt = 1:size(obj.TileTypes, 1)
                if is_turning(obj.TileTypes(tt, :))
                    obj.Concs(tt) = obj.Concs(tt) * t;
                end
            end
        end

        function apply_direction_bias(obj, d)
            % Re-weights tiles according to their output growth direction
            % d: indices correspond to [NL NR EU ED SR SL WD WU]
            % Does not affect branching tiles
            for tt = 1:size(obj.TileTypes, 1)
                if is_direction(obj.TileTypes(tt, :), 'NL') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(1);
                elseif is_direction(obj.TileTypes(tt, :), 'NR') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(2);
                elseif is_direction(obj.TileTypes(tt, :), 'EU') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(3);
                elseif is_direction(obj.TileTypes(tt, :), 'ED') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(4);
                elseif is_direction(obj.TileTypes(tt, :), 'SR') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(5);
                elseif is_direction(obj.TileTypes(tt, :), 'SL') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(6);
                elseif is_direction(obj.TileTypes(tt, :), 'WD') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(7);
                elseif is_direction(obj.TileTypes(tt, :), 'WU') && ~strcmp(eval_tile_defn(obj.TileTypes(tt, :)), "branching")
                    obj.Concs(tt) = obj.Concs(tt) * d(8);
                end
            end
        end

        function apply_turning_angle(obj, angle)
            % Warning: this implementation is deprecated but remains as a
            % placeholder
            % Re-weights tile types according to their turn angle
            % Only selects for tiles with the input angle
            for tt = 1:size(obj.TileTypes, 1)
                if has_turn_angle(obj.TileTypes(tt, :), angle)
                else
                    obj.Concs(tt) = 0;
                end
            end
        end
        
        %% Tile getters
        function defn = get_weighted_random(obj)
            cdf = [0 zeros(1, size(obj.Concs, 1))];
            for i = 1:size(obj.Concs, 1)
                cdf(i+1) = sum(obj.Concs(1:i));
            end

            % Make a random selection using cmumlative distribution
            % function and RV X
            X = rand;
            for i = 2:size(cdf, 2)
                if X < cdf(i) && X > cdf(i-1)
                    defn = obj.TileTypes(i-1, :);
                    return;
                end
            end
        end

        function defn = get_weighted_filtered_random(obj, gluename)
            % Get a valid tile type complement of parameter glue name from
            % a weighted random selection of all tile types in the tile set
            [filtered_list, filtered_concs] = obj.filter_by_input(gluename);
            % If using tile concentration kinetics, re-weight using Hill,
            % n=1
            if obj.usetileconc
                weights = filtered_concs./ (obj.Kd + filtered_concs);
            % Otherwise just re-weight linearly
            else
                weights = filtered_concs/sum(filtered_concs);
            end
            cdf = [0 zeros(1, size(weights, 1))];
            for i = 1:size(weights, 1)
                cdf(i+1) = sum(weights(1:i));
            end

            % Make a random selection using cmumlative distribution
            % function and RV X
            % Scale RV X to range of cdf
            X = min(cdf) + rand * ( max(cdf)-min(cdf) );
            for i = 2:size(cdf, 2)
                if X < cdf(i) && X > cdf(i-1)
                    defn = filtered_list(i-1, :);
                    % If Tile Concentration is enabled, re-weight for 
                    % binding probability assuming
                    % Michaelis-Menten kinetics
                    if obj.usetileconc
                        tile_conc = filtered_concs(i-1);
                        pBinding = tile_conc/(obj.Kd + tile_conc);
                        Y = rand;
                        if Y<pBinding
                            % Returns the selected tile definition that is
                            % within the range of receptor-ligand binding
                            % as defined by Michaelis-Menten kinetics
                            return;
                        else
                            % No tile should be grown
                            defn = "NULL";
                            return;
                        end
                    else
                        % Return a linearly weighted random tile
                        return;
                    end
                end
            end
            % No valid tiles exist, return NULL
            defn = "NULL";            
        end

        function [new_list, new_concs] = filter_by_input(obj, gluename)
            % Return only tile types that have the corresponding input glue
            % to the output glue parameter (gluename)
            new_list = [];
            new_concs = [];
            for tt = 1:size(obj.TileTypes, 1)
                if obj.TileTypes(tt, 1) == gluename
                    new_list = [new_list; obj.TileTypes(tt, :)];
                    new_concs = [new_concs; obj.Concs(tt)];
                end
            end
        end
    end

    methods(Static)

        %% Tile set initialization for different modes
        function [tile_list, concs] = init_linear_tile_set(concentration)
            % For the linear tile set
            % The only glue directions are WU, WD, EU, and ED
            % Only WU and WD can be inputs
            % Only EU and ED can be outputs
            % Populate all combinations of single inputs to one or two
            % outputs
            % Then double it for both the forward and backward channels

            % Initiate storage variables for the list of tile types and
            % their concentrations
            tile_list = [];
            concs = [];
            % Define the allowable glue directions
            input_dirs = ["WU", "WD"];
            output_dirs = ["EU", "ED"];

            % Enumerate combinations of channel, i/o, and directions
            fwd_input_glues = enum_glues('F', 'I', input_dirs);
            fwd_output_glues = enum_glues('F', 'O', output_dirs);
            bwd_input_glues = enum_glues('B', 'I', input_dirs);
            bwd_output_glues = enum_glues('B', 'O', output_dirs);

            % Match glues together into a standard tile type
            for i = 1:size(fwd_input_glues, 2)
                for o = 1:size(fwd_output_glues, 2)
                    g1 = fwd_input_glues(i);
                    g2 = bwd_input_glues(i);
                    g3 = fwd_output_glues(o);
                    g4 = "NULL";
                    g5 = bwd_output_glues(o);
                    g6 = "NULL";
                    if ~same_dir(g1, g3)
                        tile_list = [tile_list; g1 g2 g3 g4 g5 g6];
                        concs = [concs; concentration];
                    end
                end
            end
            
            % Match glues together into a branching tile type
            for i = 1:size(fwd_input_glues, 2)
                g1 = fwd_input_glues(i);
                g2 = bwd_input_glues(i);
                g3 = fwd_output_glues(1);
                g4 = fwd_output_glues(2);
                g5 = bwd_output_glues(1);
                g6 = bwd_output_glues(2);
                tile_list = [tile_list; g1 g2 g3 g4 g5 g6];
                concs = [concs; concentration];
            end
        end

        function [tile_list, concs] = init_radial_tile_set(concentration)
            % Initiate storage variables for the list of tile types and
            % their concentrations
            tile_list = [];
            concs = [];
            % Define the allowable glue directions
            input_dirs = ["NL", "NR", "SR", "SL", "WU", "WD", "EU", "ED"];
            output_dirs = ["NL", "NR", "SR", "SL", "WU", "WD", "EU", "ED"];

            % Enumerate combinations of channel, i/o, and directions
            fwd_input_glues = enum_glues('F', 'I', input_dirs);
            fwd_output_glues = enum_glues('F', 'O', output_dirs);
            bwd_input_glues = enum_glues('B', 'I', input_dirs);
            bwd_output_glues = enum_glues('B', 'O', output_dirs);

            % Radial needs to enforce additional rules:
            % 1. Since all directions can be inputs AND outputs, output
            % glues cannot be in the same direction as the inputs
            % 2. On branching tiles, output glues need to be on the same
            % side as each other
            % 3. Corner glues (e.g. input on SR, output on ED) are not
            % allowed as it is an immediate conflict

            % Match glues together into a standard tile type
            for i = 1:size(fwd_input_glues, 2)
                for o = 1:size(fwd_output_glues, 2)
                    g1 = fwd_input_glues(i);
                    g2 = bwd_input_glues(i);
                    g3 = fwd_output_glues(o);
                    g4 = "NULL";
                    g5 = bwd_output_glues(o);
                    g6 = "NULL";
                    if ~same_side(g1, g3) && ~same_corner(g1, g3)
                        tile_list = [tile_list; g1 g2 g3 g4 g5 g6];
                        concs = [concs; concentration];
                    end
                end
            end
            
            % Match glues together into a branching tile type
            fwd_combinations = combinations(fwd_output_glues, fwd_output_glues);
            bwd_combinations = combinations(bwd_output_glues, bwd_output_glues);

            for i = 1:size(fwd_input_glues, 2)
                for o = 1:size(fwd_combinations, 1)
                    g1 = fwd_input_glues(i);
                    g2 = bwd_input_glues(i);
                    g3 = fwd_combinations(o, 1);
                    g4 = fwd_combinations(o, 2);
                    g5 = bwd_combinations(o, 1);
                    g6 = bwd_combinations(o, 2);
                    if ~same_side(g1, g3) && ~same_side(g1, g4) && same_side(g3, g4) && ~same_dir(g3, g4) && ~same_corner(g1, g3) && ~same_corner(g1, g4)
                        tile_list = [tile_list; g1 g2 g3 g4 g5 g6];
                        concs = [concs; concentration];
                    end
                end
            end
        end
    end
end

File: ystam/tile/tileset.py
"""TileSet: manages tile type population with weighted random selection."""

from __future__ import annotations

import numpy as np

from ystam.helpers import combinations
from ystam.tile.glue import (
    Glue,
    NULL,
    enum_glues,
    eval_tile_defn,
    is_direction,
    is_turning,
    same_corner,
    same_dir,
    same_side,
)

# 8 radial directions
RADIAL_DIRS = ["NL", "NR", "SR", "SL", "WU", "WD", "EU", "ED"]


class TileSet:
    """Manages tile types and their concentrations for weighted random selection.

    A tile definition is a 6-element list:
        [fwd_in, bwd_in, fwd_out1, fwd_out2, bwd_out1, bwd_out2]
    where fwd_out2/bwd_out2 are None for standard (non-branching) tiles.
    """

    def __init__(
        self,
        mode: str = "radial",
        use_tile_conc: bool = False,
        kd: float = 1.0,
        conc_max: float = 100.0,
        rng: np.random.Generator | None = None,
    ) -> None:
        self.kd = kd
        self.conc_max = conc_max
        self.use_tile_conc = use_tile_conc
        self.rng = rng or np.random.default_rng()

        if mode == "radial":
            self.tile_types, self.concs = self._init_radial_tile_set(conc_max)
        else:
            raise ValueError(f"Unsupported mode: {mode}")

    # ------------------------------------------------------------------
    # Concentration modifiers
    # ------------------------------------------------------------------

    def dec_conc(self, defn: list[Glue | None]) -> None:
        """Decrease concentration of matching tile type by 1."""
        idx = self._find_defn_index(defn)
        if idx is not None and self.concs[idx] > 0:
            self.concs[idx] -= 1

    def inc_conc(self, defn: list[Glue | None]) -> None:
        """Increase concentration of matching tile type by 1."""
        idx = self._find_defn_index(defn)
        if idx is not None:
            self.concs[idx] += 1

    def buffer(self, num: float) -> None:
        """Increase all tile concentrations by num."""
        self.concs = [c + num for c in self.concs]

    def apply_branching_factor(self, b: float) -> None:
        """Scale concentration of branching tiles by factor b."""
        for i, defn in enumerate(self.tile_types):
            if defn[3] is not None and defn[5] is not None:
                self.concs[i] *= b

    def apply_turning_factor(self, t: float) -> None:
        """Scale concentration of turning tiles by factor t."""
        for i, defn in enumerate(self.tile_types):
            if is_turning(defn):
                self.concs[i] *= t

    def apply_direction_bias(self, d: list[float]) -> None:
        """Scale concentrations by output direction.

        d indices: [NL, NR, EU, ED, SR, SL, WD, WU].
        """
        dir_map = {
            "NL": 0, "NR": 1, "EU": 2, "ED": 3,
            "SR": 4, "SL": 5, "WD": 6, "WU": 7,
        }
        for i, defn in enumerate(self.tile_types):
            if eval_tile_defn(defn) == "branching":
                continue
            out = defn[2]
            if out is not None:
                direction = out.direction
                if direction in dir_map:
                    self.concs[i] *= d[dir_map[direction]]

    # ------------------------------------------------------------------
    # Random tile selection
    # ------------------------------------------------------------------

    def get_weighted_random(self) -> list[Glue | None]:
        """Select a random tile type weighted by concentration."""
        total = sum(self.concs)
        if total <= 0:
            return [None] * 6
        probs = [c / total for c in self.concs]
        idx = self.rng.choice(len(self.tile_types), p=probs)
        return list(self.tile_types[idx])

    def get_weighted_filtered_random(self, input_glue: Glue) -> list[Glue | None] | None:
        """Select a random compatible tile type that accepts the given input glue.

        Returns the tile definition or None if no valid tile exists / binding check fails.
        """
        filtered, filtered_concs = self._filter_by_input(input_glue)
        if not filtered:
            return None

        # Weight by Michaelis-Menten or linear
        conc_arr = np.array(filtered_concs, dtype=float)
        if self.use_tile_conc:
            weights = conc_arr / (self.kd + conc_arr)
        else:
            total = conc_arr.sum()
            if total <= 0:
                return None
            weights = conc_arr / total

        total_w = weights.sum()
        if total_w <= 0:
            return None

        # CDF-based selection matching MATLAB behavior
        cdf = np.concatenate(([0.0], np.cumsum(weights)))
        x = cdf[0] + self.rng.random() * (cdf[-1] - cdf[0])
        selected_idx = 0
        for j in range(1, len(cdf)):
            if cdf[j - 1] < x <= cdf[j]:
                selected_idx = j - 1
                break

        defn = list(filtered[selected_idx])

        # If tile concentration enabled, apply binding probability check
        if self.use_tile_conc:
            tile_conc = filtered_concs[selected_idx]
            p_binding = tile_conc / (self.kd + tile_conc)
            if self.rng.random() >= p_binding:
                return None

        return defn

    # ------------------------------------------------------------------
    # Tile set initialization
    # ------------------------------------------------------------------

    @staticmethod
    def _init_radial_tile_set(
        concentration: float,
    ) -> tuple[list[list[Glue | None]], list[float]]:
        """Generate all valid radial tile types with their concentrations."""
        tile_list: list[list[Glue | None]] = []
        concs: list[float] = []

        fwd_in = enum_glues("F", "I", RADIAL_DIRS)
        fwd_out = enum_glues("F", "O", RADIAL_DIRS)
        bwd_in = enum_glues("B", "I", RADIAL_DIRS)
        bwd_out = enum_glues("B", "O", RADIAL_DIRS)

        # Standard tiles: 1 input -> 1 output
        for i, fi in enumerate(fwd_in):
            for o, fo in enumerate(fwd_out):
                if not same_side(fi, fo) and not same_corner(fi, fo):
                    tile_list.append([fi, bwd_in[i], fo, None, bwd_out[o], None])
                    concs.append(concentration)

        # Branching tiles: 1 input -> 2 outputs
        fwd_combos = combinations(fwd_out, fwd_out)
        bwd_combos = combinations(bwd_out, bwd_out)

        for i, fi in enumerate(fwd_in):
            for (fo1, fo2), (bo1, bo2) in zip(fwd_combos, bwd_combos):
                if (
                    not same_side(fi, fo1)
                    and not same_side(fi, fo2)
                    and same_side(fo1, fo2)
                    and not same_dir(fo1, fo2)
                    and not same_corner(fi, fo1)
                    and not same_corner(fi, fo2)
                ):
                    tile_list.append([fi, bwd_in[i], fo1, fo2, bo1, bo2])
                    concs.append(concentration)

        return tile_list, concs

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _filter_by_input(
        self, input_glue: Glue
    ) -> tuple[list[list[Glue | None]], list[float]]:
        """Return tile types whose first input matches the given glue."""
        filtered = []
        filtered_concs = []
        for defn, conc in zip(self.tile_types, self.concs):
            if defn[0] is not None and defn[0] == input_glue:
                filtered.append(defn)
                filtered_concs.append(conc)
        return filtered, filtered_concs

    def _find_defn_index(self, defn: list[Glue | None]) -> int | None:
        """Find index of a tile definition in the tile set."""
        for i, td in enumerate(self.tile_types):
            if td == defn:
                return i
        return None
